### PR TITLE
ARROW-8720: [C++] Fix checked_pointer_cast ifdef logic

### PR DESCRIPTION
**Pull request by fsaintjacques**
The expensive version (dynamic_pointer_cast) was enabled in release build instead of debug builds. This aligns with `checked_cast`.

### CODE REVIEW

**Comment by pitrou** [comment]
`NDEBUG` is still as confusing as it was 30 years ago.

**Review by pitrou — APPROVED** [review]
+1. It's also better for debug builds :-)

**Comment by fsaintjacques** [comment]
Our brains are stack machines with a depth of 1, it's hard to parse not-not-debug.
